ired.1
.Dd Oct 7, 2009
.Dt IRED 1
.Os
.Th IRED 1
.Sh NAME
ired \- interactive raw editor
.Sh SYNOPSIS
.Nm ired
.Op Fl hnv
.Op Fl i Ar script
.Ar file ...
.Sh DESCRIPTION
Simple and minimal hexadecimal editor based on radare concepts
.Pp
The program arguments are:
.Pp
.Bl -tag -width Fl
.It Fl h
Show help message
.It Fl v
Show version information
.It Fl n
Do not be verbose
.It Fl i Ar script
Interpret a script file as ired commands.
.Pp
Same as "ired file < script"
.El
.PP
.Sh COMMANDS
.Pp
.Bl -tag -width Fl
.It s[+-addr] 
seek to relative or absolute address
.Pp
> s 0x400  # seek to offset 0x400
.Pp
> s+20     # seek 20 bytes forward
.It b[+-size] 
change block size
.Pp
> b 5      # set block size to 5 bytes
.Pp
> b +128   # increase block size in 128 bytes
.It w[hex|"str"]
write hexpair or string
.Pp
> w 001122       # write 0x00 0x11 0x22
.Pp
> w"Hello World" # write "Hello World" string without quotes
.It /[hex|"str"]
search hexpair or string
.Pp
> /894722       # search bytes 0x89 0x47 0x22
.Pp
> /"lib"        # search string 'lib'
.It x[size]
hexdump
.Pp
> x 256    # dump 256 bytes in hexadecimal
.It X[size]
hexpair dump
.Pp
> X 0x100  # dump 256 bytes in hexpairs
.It p[fmt]
print formatted current block ('p' for help)
.Pp
ob/wW/dD/qQ  byte (oct,hex), word, dword, qword (lil, big endian)
.Pp
i/I/f/F      int32 (lil, big), float (lil, big)
.Pp
s/S          short int16 (lil, big)
.Pp
z/Z          zero-terminatted string (ascii, wide-ascii)
.Pp
./:/*        skip 1 or 4 chars, repeat last format instead of cycle
.It r[-[num]]
truncate or -remove N bytes
.Pp
> r 1024   # resize file to 1024 bytes
.Pp
> r-10     # remove 10 bytes from current seek and shrink file
.It .[file]
interpret file
.Pp
> .script # run file 'script'
.It <[file]
load file in current seek
.Pp
> <curblock.bin  # load blocksize bytes from file
.It >[file]
dump current block to file
.Pp
> >curblock.bin  # dump current block to a file
.It !cmd
run shell command
.Pp
The following environment variables are defined when found in the command string:
.Pp
FILE is the filename of the current working file
.Pp
BLOCK points to the temporal filename with the current block dump
.Pp
OFFSET and XOFFSET are decimal and hexadecimal values of the current seek
.Pp
> !cp $BLOCK $FILE-$XOFFSET.curblock # save current block
.It ?expr
calculate numeric expression
.Pp
> ?1024
.Pp
0x400 1024 02000

.It q
quit

.El
.Sh SEE ALSO
.Pp
.Xr vired(1)
.Sh AUTHORS
.PP
pancake <@nopcode.org>


ChangeLog
== release 0.2

changeset:   41:6fd885d7d0c2
tag:         tip
user:        pancake@localhost.localdomain
date:        Tue Sep 01 22:30:01 2009 +0000
files:       ChangeLog Makefile cmd.c ired.1 vired vired.1
description:
* Add manpages for ired and vired
* Compile with debugging information by default
* Added 'make dist' target
* Fix double free in cmd.c
* Fix disassemble mode in 'p' key of vired


changeset:   40:97b4743c4dc6
user:        pancake@localhost.localdomain
date:        Tue Sep 01 21:17:05 2009 +0000
files:       Makefile ired.c red.c
description:
* Rename from red to ired to avoid program name collisions


changeset:   39:03da6e158902
user:        pancake@flubox
date:        Fri Jul 17 11:47:55 2009 +0200
files:       Makefile cmd.c
description:
* Fix search command. Oops :)


changeset:   38:1715fc1d195e
user:        pancake@localhost.localdomain
date:        Sun Jul 12 01:22:15 2009 +0000
files:       cmd.c
description:
* Fix segfault in cmd_search()


changeset:   37:8ff91ac81732
user:        pancake@localhost.localdomain
date:        Sat Jul 11 16:31:06 2009 +0000
files:       ChangeLog red.h
description:
* Bump revision number to 0.2


== release 0.1

changeset:   36:ba4108a3c6d5
tag:         tip
user:        pancake@localhost.localdomain
date:        Sat Jul 11 16:28:35 2009 +0000
files:       cmd.c
description:
* Fix gcc warning. Release 0.1


changeset:   35:a69c81ce6b79
user:        pancake@localhost.localdomain
date:        Sat Jul 11 16:15:18 2009 +0000
files:       Makefile
description:
* Use CFLAGS when compiling


changeset:   34:d3c02ddef876
user:        pancake@localhost.localdomain
date:        Sat Jul 11 16:14:37 2009 +0000
files:       cmd.c
description:
* Fix '!' command segfault
  - Initialize buf in search command


changeset:   33:cd894a746246
user:        pancake@localhost.localdomain
date:        Sat Jul 11 13:58:12 2009 +0000
files:       util.c
description:
* Use 'isascii' instead of buf[0] for pz and pZ


changeset:   32:795733b782ee
user:        pancake@localhost.localdomain
date:        Sat Jul 11 13:41:51 2009 +0000
files:       io.c red.h
description:
* Full implementation of the w32 IO API


changeset:   31:af7a10ccfed7
user:        pancake@localhost.localdomain
date:        Sat Jul 11 13:20:11 2009 +0000
files:       Makefile red.h
description:
* First build for w32
  - Added 'make w32'
  - TODO: Use w32 IO API instead of posix by default on w32


changeset:   30:23d74ef2b61b
user:        pancake@localhost.localdomain
date:        Sat Jul 11 13:09:01 2009 +0000
files:       README
description:
* Some minor changes in README


changeset:   29:f46469969239
user:        pancake@localhost.localdomain
date:        Sat Jul 11 13:02:59 2009 +0000
files:       Makefile util.c
description:
* Added support for 'K' 'M' and 'G' suffix in str2ull function
  - Added support for octal input (start with '0')
  - po command shows in the same octal way ('0' prefix)
* Fix hexstr2bin with spaces
* Do not make loc vired


changeset:   28:bdcc2e8cfb90
user:        pancake@localhost.localdomain
date:        Sat Jul 11 10:58:28 2009 +0000
files:       cmd.c
description:
* Some code review and simplification -10LOC


changeset:   27:b944c1a84437
user:        pancake@localhost.localdomain
date:        Sat Jul 11 10:54:29 2009 +0000
files:       README red.c
description:
* Minor fix in README
* Remove end '.' in error message


changeset:   26:26582b52ed09
user:        pancake@localhost.localdomain
date:        Sat Jul 11 10:52:13 2009 +0000
files:       util.c
description:
* Fix build


changeset:   25:05b4d6b61a48
user:        pancake@localhost.localdomain
date:        Sat Jul 11 10:50:22 2009 +0000
files:       README
description:
* Fix README


changeset:   24:ef1a2671bbc6
user:        pancake@localhost.localdomain
date:        Sat Jul 11 10:47:11 2009 +0000
files:       cmd.c io.c red.c util.c
description:
* hexdump() now gets a 3rd argument to set the width
* Show filesize when no args given to 'r' command
  - Truncate file when removing bytes (f.ex: r-10)
* Fix segfault in cmd_system() when no BLOCK is used, but [X]OFFSET is
* skipspaces() is now skipping 'all' spaces, not only the first one
* Some code cleanup (-5 LOC) - now is 499 LOC :P


changeset:   23:4f672b5b67b9
user:        pancake@localhost.localdomain
date:        Sat Jul 11 01:26:54 2009 +0000
files:       cmd.c vired
description:
* Fix segfault coz an uninitialized variable
* 'r' with no argument should return filesize
  - not implemented


changeset:   22:0b6a9d1bcae2
user:        pancake@localhost.localdomain
date:        Sat Jul 11 01:20:51 2009 +0000
files:       Makefile cmd.c io.c red.c vired
description:
* Add command 'r' to resize and remove bytes
  r128  : truncate file to 128 bytes
  r-10  : remove 10 bytes from current seek
* Added io_truncate
* warning: 500LOC reached :)


changeset:   21:1de30455d1c5
user:        pancake@localhost.localdomain
date:        Sat Jul 11 00:34:57 2009 +0000
files:       red.c
description:
* Simplify argv parsing (-1 LOC)


changeset:   20:d37eb8dbb125
user:        pancake@localhost.localdomain
date:        Sat Jul 11 00:27:02 2009 +0000
files:       cmd.c red.c util.c
description:
* Add '?' in help message
* Simplify and add more checks in cmd_system()
  - setenv FILE once
* Simplify code for parsing arguments
  - Minor simplification in hexstr2raw


changeset:   19:f455f21e48f9
user:        pancake@localhost.localdomain
date:        Sat Jul 11 00:07:45 2009 +0000
files:       Makefile cmd.c red.c util.c
description:
* Code cleanup (-18 LOC)
  - Move print_fmt stuff into utils.c
* Use static in more functions


changeset:   18:de171facddfb
user:        pancake@localhost.localdomain
date:        Fri Jul 10 23:43:24 2009 +0000
files:       Makefile cmd.c
description:
* Fix some valgrind warnings in the print format command
  - Some uninitialized variable usage
  - Out of buffer reads are now controlled
  - repeat last print format until the end of the block


ired.c
/* Copyleft 2009 -- pancake /at/ nopcode /dot/ org */

#define ut64 unsigned long long 
#define ut8 unsigned char

static int verbose = 1;
static char *script = 0;
static ut64 oldseek, curseek = 0LL;
static int obsize, bsize = 256;
static int red_cmd(char *cmd); // XXX : recursive depenency

#include "red.h"
#include "util.c"
#include "cmd.c"

static void red_slurpin() {
	ut8 buf[4096];
	for(;;) {
		int len = read(0, buf, 4096);
		if (len<1) break;
		hexdump(buf, len, 16);
		curseek += len;
	}
}

static void red_interpret(char *file) {
	char buf[1024];
	FILE *fd = fopen(file, "r");
	if (fd != NULL) {
		for(;;) {
			if (fgets(buf, 1023, fd) == NULL)
				break;
			red_cmd(buf);
		}
		fclose(fd);
	} else fprintf(stderr, "Cannot open script file '%s'\n", file);
}

static int red_cmd(char *cmd) {
	switch(*cmd) {
	case 'q': return 0;
	case ';': case '#': break; // comment
	case '>': cmd_dump(cmd+1); break;
	case '<': cmd_load(cmd+1); break;
	case '.': red_interpret(skipspaces(cmd+1)); break;
	case 's': cmd_seek(cmd+1); break;
	case 'b': cmd_bsize(cmd+1); break;
	case '/': cmd_search(cmd+1); break;
	case 'p': cmd_print(cmd+1); break;
	case 'r': cmd_resize(cmd+1); break;
	case 'x': cmd_hexdump(cmd+1); break;
	case 'X': cmd_bytedump(cmd+1); break;
	case 'w': cmd_write(cmd+1); break;
	case '!': cmd_system(cmd+1); break;
	case '?': cmd_help(cmd+1); break;
	default: fprintf(stderr, "? %s\n", cmd);
	}
	return 1;
}

static int red_prompt() {
	char *at, *at2, line[4096];
	if (verbose) {
		printf("[0x%08"LLF"x]> ", curseek);
		fflush(stdout);
	}
	if (fgets(line, 4095, stdin) == NULL)
		return 0;
	line[strlen(line)-1] = '\0';
	if (*line != '!') {
		at = strchr(line, '@');
		oldseek = curseek;
		obsize = bsize;
		if (at) {
			*at = 0;
			at2 = strchr(++at, ':');
			if (at2) {
				*at2 = 0; at2++;
				if (*at2) bsize = (int)str2ut64(at2);
			}
			if (*at) curseek = str2ut64(at);
		}
	}
	return red_cmd(skipspaces(line));
}

static int red_open(char *file) {
	int ret = io_open(file);
	if (ret != -1) {
		oldseek = 0;
		setenv("FILE", file, 1);
		if (script)
			red_interpret(script);
		while(red_prompt()) {
			curseek = oldseek;
			bsize = obsize;
		}
		io_close();
	} else fprintf(stderr, "Cannot open '%s'\n", file);
	return ret==-1 ?1:0;;
}

static int red_help() {
	puts("ired [-hnv] [-i script] [file] [..]");
	return 0;
}

int main(int argc, char **argv) {
	int i, ret = 1;
	if (argc>1)
	for(i=1;i<argc;i++) {
		if (argv[i][0]=='-')
			switch(argv[i][1]) {
			case 'i': script = argv[++i]; break;
			case 'n': verbose = 0; break;
			case 'v': puts("red "VERSION" 2009"); ret = 0; break;
			case 'h': ret = red_help(); break;
			case 0x0: red_slurpin(); ret = 0; break;
		} else ret = red_open(argv[i]);
	} else ret = red_help();
        return ret;
}
